services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticatorFactory.java

```diff
@@ -0,0 +1,109 @@
+package org.keycloak.authentication.authenticators.sessionlimits;
+
+import org.keycloak.Config;
+import org.keycloak.authentication.Authenticator;
+import org.keycloak.authentication.AuthenticatorFactory;
+import org.keycloak.models.AuthenticationExecutionModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.provider.ProviderConfigProperty;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class UserSessionLimitsAuthenticatorFactory implements AuthenticatorFactory {
+    public static final String USER_REALM_LIMIT = "userRealmLimit";
+    public static final String USER_CLIENT_LIMIT = "userClientLimit";
+    public static final String BEHAVIOR = "behavior";
+    public static final String DENY_NEW_SESSION = "Deny new session";
+    public static final String TERMINATE_OLDEST_SESSION = "Terminate oldest session";
+    public static final String USER_SESSION_LIMITS = "user-session-limits";
+    public static final String ERROR_MESSAGE = "errorMessage";
+
+    private static AuthenticationExecutionModel.Requirement[] REQUIREMENT_CHOICES = {
+            AuthenticationExecutionModel.Requirement.REQUIRED,
+            AuthenticationExecutionModel.Requirement.DISABLED
+    };
+
+    @Override
+    public String getDisplayType() {
+        return "User session count limiter";
+    }
+
+    @Override
+    public String getReferenceCategory() {
+        return null;
+    }
+
+    @Override
+    public boolean isConfigurable() {
+        return true;
+    }
+
+    @Override
+    public AuthenticationExecutionModel.Requirement[] getRequirementChoices() {
+        return REQUIREMENT_CHOICES;
+    }
+
+    @Override
+    public boolean isUserSetupAllowed() {
+        return false;
+    }
+
+    @Override
+    public String getHelpText() {
+        return "Configures how many concurrent sessions a single user is allowed to create for this realm and/or client";
+    }
+
+    @Override
+    public List<ProviderConfigProperty> getConfigProperties() {
+        ProviderConfigProperty userRealmLimit = new ProviderConfigProperty();
+        userRealmLimit.setName(USER_REALM_LIMIT);
+        userRealmLimit.setLabel("Maximum concurrent sessions for each user");
```

Comment:
One additional point: IMO it will be nice if USER_CLIENT_LIMIT is disabled by default (hence having 0 or negative value). Not sure what is the best default value for USER_REALM_LIMIT, but maybe it will be nice to add some default value. How about 3? I am leaving to you what you consider to be best default...

Reply:
Done as requested; the tooltips now mention how to disable the limit. 
USER_CLIENT_LIMIT is disabled by default and USER_REALM_LIMIT is set to 3. This seems a sensible default to me.